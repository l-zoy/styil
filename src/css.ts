import { StyleSheetOptions } from './types'
import { createSelector } from './createSelector'
import { StyleSheet } from './sheet'
import { parseRules } from './parse'
import { CSSAttribute } from 'nativeCssTypes'

export function createStyil(options: Partial<StyleSheetOptions> = {}) {
  const { key = 'css', container, speedy, nonce } = options

  const globalCache = new Set<string>([])

  const sheet = new StyleSheet({
    key,
    speedy: speedy === undefined ? process.env.NODE_ENV === 'production' : speedy,
    container: globalThis.document ? container ?? globalThis.document.head : null,
    nonce
  })

  function createStyle(options: { styles: CSSAttribute; keyframes?: boolean; glob?: boolean }) {
    const { styles, keyframes, glob } = options
    const className = `${key}-${createSelector(styles)}`

    if (!globalCache.has(className)) {
      globalCache.add(className)

      sheet.insertStyle(
        parseRules(
          keyframes ? { [`@keyframes ${className}`]: styles } : styles,
          glob ? undefined : `.${className}`
        )
      )
    }

    return glob ? undefined : className
  }

  function css(styles: CSSAttribute) {
    return createStyle({ styles })
  }

  function glob(styles: CSSAttribute) {
    return createStyle({ styles, glob: true })
  }

  function keyframes(styles: CSSAttribute) {
    return createStyle({ styles, keyframes: true })
  }

  function getCssValue() {
    return sheet.ssrData
  }

  function flush() {
    sheet.flush()
    globalCache.clear()
  }

  return { css, glob, keyframes, getCssValue, flush }
}

export const { css, glob, keyframes, getCssValue, flush } = createStyil()
