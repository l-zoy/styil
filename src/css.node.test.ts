/**
 * @jest-environment node
 */

import { createStyils, getCssValue as defaultGetCssValue, css as defaultCss } from './css'

describe('css node', () => {
  it('createStyils node', () => {
    const { getCssValue, css } = createStyils()

    css({
      width: 100
    })

    css({
      width: 100
    })

    css({
      height: 100
    })

    expect(getCssValue().join('')).toMatchSnapshot()
  })

  it('node default getCssValue ', () => {
    defaultCss({
      width: 100
    })

    defaultCss({
      width: 100
    })

    defaultCss({
      height: 100
    })

    expect(defaultGetCssValue().join('')).toMatchSnapshot()
  })
})
