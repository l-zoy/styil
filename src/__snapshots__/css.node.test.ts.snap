// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`css node createStyils node 1`] = `".css-2175997142{width:100px;}.css-2374624703{height:100px;}"`;

exports[`css node node default getCssValue  1`] = `".css-2175997142{width:100px;}.css-2374624703{height:100px;}"`;
